### PR TITLE
Move chat message into CanJoinTeam()

### PR DESCRIPTION
**Pull request by ChillerDragon**
Now `CGameContext` no longer assumes the `IGameController` declined the team join due to slots.

This enables custom gametypes to disallow joining the game if the player died, an active tournament is running or the player is not logged in yet. And then the controller can print the correct error message accordingly.

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by heinrich5991** [comment]
I think it'd be cleaner to pass a buffer to populate with the reason, unconditionally. Could you do that?

**Comment by ChillerDragon** [comment]
> I think it'd be cleaner to pass a buffer to populate with the reason, unconditionally. Could you do that?

Yes sounds good will do when I get to it.

**Comment by Kaffeine** [comment]
Oh. I came to say that this breaks encapsulation and responsibility: `CanJoinTeam()` does not sound like a talkative peace of code. With the current name I'd expect it to be called regardless of the player activity.

And then it turned out that I [did the same](https://github.com/infclass/teeworlds-infclassR/commit/e0b5b06ca68cb9b54f40c550e5b904d6aa300b1d) 😀. Okay!

**Comment by Kaffeine** [comment]
> I think it'd be cleaner to pass a buffer to populate with the reason, unconditionally. Could you do that?

TL;DR: It won't work with translations, thought in this particular method it is mostly irrelevant (excepting the demo recording).

#### Details
Infclass mod has server-side translations. It supports arguments substitution, plurals, etc:
- https://github.com/infclass/teeworlds-infclassR/blob/1a16492ac5bc5ad04fce44962d199b62c79c0838/src/teeuniverses/components/localization.h#L163
- https://github.com/Kaffeine/teeworlds-infclassR2/blob/44cb186d5d5d4659e7cf13444091f0059d537369/src/game/server/gamecontext.h#L362-L372

Translation looks like [this](https://github.com/infclass/teeworlds-infclassR/blob/1a16492ac5bc5ad04fce44962d199b62c79c0838/other/po/de/infclass.po#L1308C15-L1308C15).

The translation expects the original message before arguments substitution, and the main flow is this:
1. Game code uses `SendBroadcast_Localization(ClientID, const char* pText, ...)`
2. Formats 'for-demo' message (formatted for `"en"` / `sv_demo_lang`)
3. The localization method iterates over the receivers (all clients in case of `ClientID == -1`) and for each client:
    1. Calls `Format_VL(Buffer, m_apPlayers[i]->GetLanguage(), pText, VarArgs);` to format the translated string
    2. Constructs the net message and uses SendPackMsg() to send it to the client (`MSGFLAG_NORECORD`)
    4. Calls `SendPackMsg()` with 'for-demo' formatted text and `MSGFLAG_NOSEND` 

With the buffer formatted in `CanJoinTeam()` it won't be possible to generate `en` translated message for the demo recorder.

**Comment by Kaffeine** [comment]
> I think it'd be cleaner to pass a buffer to populate with the reason, unconditionally. Could you do that?

I don't understand the reason to do it. Do you want extra arguments like that:
`virtual bool CanJoinTeam(int Team, int NotThisID, char *pBuffer, int BufferSize);`?

What is the use-case? As you see, currently we use SendChatTarget() for the `/pause` message and `SendBroadcast()` for the `"Only %d active players ...` message. In other words, the caller has no information to choose between Chat and Broadcast.

**Comment by heinrich5991** [comment]
> I don't understand the reason to do it. Do you want extra arguments like that:
> `virtual bool CanJoinTeam(int Team, int NotThisID, char *pBuffer, int BufferSize);`?

Yes.





> What is the use-case? As you see, currently we use SendChatTarget() for the `/pause` message and `SendBroadcast()` for the `"Only %d active players ...` message. In other words, the caller has no information to choose between Chat and Broadcast.

It can either stay outside (since you can actually join the spectators when you're paused, you're just stopped because you might be clicking that button accidentally), or it can also go into broadcast I guess.

**Comment by heinrich5991** [comment]
I think that's not a good design for translations for DDNet, as it relies on server updates for localization updates when it is rather a UI thing, i.e. it concerns the client. See also #3091.

**Comment by Kaffeine** [comment]
> It can either stay outside (since you can actually join the spectators when you're paused, you're just stopped because you might be clicking that button accidentally), or it can also go into broadcast I guess.

It is the mod (DDNet) specific behavior. It makes no sense to me to extract the messages to the GameController so mods can customize them, and at the same time hardcode the message behavior.

**Comment by heinrich5991** [comment]
> It is the mod (DDNet) specific behavior.

/pause is mod-specific behavior, so are teams, I don't understand where we draw the lines yet.

**Comment by Kaffeine** [comment]
> I think that's not a good design for translations for DDNet, as it relies on server updates for localization updates when it is rather a UI thing, i.e. it concerns the client. See also #3091.

Sounds like the 0.7 mod unfriendly way. The argument "it relies on %% updates for localization updates" works in both ways. There is a number of users who use old DDNet clients on new servers, and in your case you "rely on the client updates for localization", which is both relies on more updates (there are more clients than servers, instead of one update you're asking every player to update) AND works against the mods.

**Comment by Kaffeine** [comment]
> I don't understand where we draw the lines yet.

You're removing the possibility to modify the message delivery method (chat, broadcast, motd).

**Comment by heinrich5991** [comment]
> > I don't understand where we draw the lines yet.
> 
> You're removing the possibility to modify the message delivery method (chat, broadcast, motd). And it also makes it harder (or virtually impossible) to translate the messages.

It seems okay to me to just use a single response type for a button click.

**Comment by Kaffeine** [comment]
> It seems okay to me to just use a single response type for a button click.

Hmm. Makes sense if you put it this way.

Maybe the mod needs more than one message to explain something but hm, not a big deal.

**Comment by ChillerDragon** [comment]
Oh yea hm true

``Use /pause first then you can kill`` used to be a chat message and is now a broadcast. Idk is that bad?

**Review by heinrich5991 — APPROVED** [review]
Thanks.
